### DISCUSSION TITLE
How to draw a custom texture/material in polygon (square, ...)?

### DISCUSSION THREAD

**Discussion by PepaZapletal** [sceneview/sceneform-android#287] [Q&A]
I'm trying to draw texture inside defined points (3, 4 or more points).

I found this: https://stackoverflow.com/questions/54433707/how-to-draw-a-polygon-with-sceneform-arcore and it's a good start.

Then I found this tutorial: https://catlikecoding.com/unity/tutorials/procedural-meshes/creating-a-mesh/ and updated the code from the previous post. Like this:

```
        val uvTopLeft = UvCoordinate(0.0f, 1.0f)
        val uvTopRight = UvCoordinate(1.0f, 1.0f)
        val uvBotLeft = UvCoordinate(0.0f, 0.0f)
        val uvBotRight = UvCoordinate(1.0f, 0.0f)
        val vertices: List<Vertex> = ArrayList(
            Arrays.asList(
                Vertex.builder().setPosition(p0).setNormal(up).setUvCoordinate(uvTopLeft).build(),
                Vertex.builder().setPosition(p1).setNormal(up).setUvCoordinate(uvTopRight).build(),
                Vertex.builder().setPosition(p2).setNormal(up).setUvCoordinate(uvBotLeft).build(),
                Vertex.builder().setPosition(p3).setNormal(up).setUvCoordinate(uvBotRight).build()
            )
        )
        val triangleIndices: MutableList<Int> = ArrayList(3)
        triangleIndices.add(0)
        triangleIndices.add(2)
        triangleIndices.add(1)
        triangleIndices.add(2)
        triangleIndices.add(3)
        triangleIndices.add(1)
```
Now it's draw texture inside 4 points (you have to insert points in correct order). But it's drawn as two triangles. So the common edges may not fit exactly.

Is it possible to somehow draw texture/material inside defined points in SceneView? Something like this:

![Výstřižek](https://i.stack.imgur.com/KGKAz.jpg)

Also, it's possible to rotate or zoom in/out the texture?

EDIT: Or maybe draw it as a model/object with texture?

**Comment by grassydragon** [comment]
Hi!

> Is it possible to somehow draw texture/material inside defined points in SceneView?

Yes, if you are drawing a plane, you can just use the X and Z world coordinates of the points as the UV coordinates.

> Also, it's possible to rotate or zoom in/out the texture?

Yes, you can apply rotation and scale to the UV coordinates.

> Or maybe draw it as a model/object with texture?

In the Stack Overflow tutorial that you referenced the plane is created as an object. It may be just more difficult to apply a texture to more complex objects.

**Comment by PepaZapletal** [comment]
Hi @grassydragon ,

Thank you a lot for your answer!

> Yes, if you are drawing a plane, you can just use the X and Z world coordinates of the points as the UV coordinates.

Yes, I'm drawing only a plane. Can you please write me more about this solution? I have `ArrayList<AnchorNode>` of points.  So I have for every point `worldPosition` (X and Z world coordinates)

But how can I transform it to the UV coordinates? And then draw it on the surface?

I'm trying to find more info about it but documentation about ARCore/SceneForm is ... isn't :)

Can you please write me some more info? Where should I start? What methods to use?

**Comment by grassydragon** [comment]
> But how can I transform it to the UV coordinates?

You can first try using them as is because the texture will repeat when you go out of bounds.

> Can you please write me some more info? Where should I start? What methods to use?

You need to generate triangles using your points. For example, `PlaneVisualizer` generates triangles for a plane detected by ARCore, however, the plane points form a convex polygon as I remember and the approach can be different for concave polygons:
https://github.com/SceneView/sceneform-android/blob/1476ddeaa26ac7a28536981a1bf6222bb44c85e6/core/src/main/java/com/google/ar/sceneform/rendering/PlaneVisualizer.java#L201

The main goal is to find a way to generate triangles to render your polygon and the other things should be similar to the Stack Overflow code.

**Comment by PepaZapletal** [comment]
Thank you. I will look at it.

One more question: **what if we limit the points to 4? (user define 4 points/anchors)**

It will be 4 vertex/vertices of the polygon. Then we will draw this polygon/object. But with depth (height) of that object will be small (a few pixels) so it will look like a plane.

And then we can draw texture on it.

What do you think about it? Is it possible? Maybe it will be easier than the solution with generating triangles to render a polygon. 

Is it possible in SceneView to render polygon defined by 4 vertexes? 

@grassydragon

**Reply by grassydragon** [reply]
> What do you think about it? Is it possible? Maybe it will be easier than the solution with generating triangles to render a polygon.

I think it isn't possible since the user defines arbitrary points and we can't modify a static object.

> Is it possible in SceneView to render polygon defined by 4 vertexes?

No, there are only a cube, sphere and cylinder.

**Reply by ThomasGorisse** [reply]
> Is it possible in SceneView to render polygon defined by 4 vertexes?

You have to use your own triangulation system.
I personnaly made it with [Delaunay Triangulation](https://en.wikipedia.org/wiki/Delaunay_triangulation) but can't share the code freely.

**Reply by PepaZapletal** [reply]
Thank you both for the info!

@ThomasGorisse Can I buy the code from you? (Pay it as sponsorship for example.)

I will send you an email about it ok? Thank you.
